tqdm/_tqdm.py

```diff
@@ -41,12 +43,75 @@ def __init__(self, msg, fp_write=None, *a, **k):
             super(TqdmDeprecationWarning, self).__init__(msg, *a, **k)
 
 
+class TMonitor(Thread):
+    """ tqdm monitoring thread
+    Monitors if tqdm bars are taking too much time to display
+    and readjusts miniters automatically if necessary"""
+
+    # internal vars for unit testing
+    _time = None
+    _sleep = None
+
+    def __init__(self, tqdm_cls, sleep_interval):
+        Thread.__init__(self)
+        self.exit_event = Event()
+        self.daemon = True  # kill thread when main killed (KeyboardInterrupt)
+        self.was_killed = False
+        self.tqdm_cls = tqdm_cls
+        self.sleep_interval = sleep_interval
+        if TMonitor._time is not None:
+            self._time = TMonitor._time
+        else:
+            self._time = time
+        if TMonitor._sleep is not None:
+            self._sleep = TMonitor._sleep
+        else:
+            self._sleep = sleep
+        self.start()
+
+    def exit(self):
+        self.exit_event.set()
+        self.was_killed = True
+        # self.join()  # DO NOT, blocking event, slows down tqdm at closing
+        return self.report()
+
+    def run(self):
+        while True:
+            # Sleep some time...
+            self._sleep(self.sleep_interval)
+            # Quit if killed
+            # if self.exit_event.isSet():  # TODO: should work but does not...
+            if self.was_killed:
+                return
+            # Then monitor!
+            cur_t = self._time()
+            # Check for each tqdm instance if one is waiting too long to print
+            for instance in self.tqdm_cls._instances:
+                # Only if mininterval > 1 (else iterations are just slow)
+                # and last refresh was longer than maxinterval in this instance
+                if instance.miniters > 1 and \
+                  (cur_t - instance.last_print_t) >= instance.maxinterval:
+                    # We force bypassing miniters on next iteration
+                    # dynamic_miniters should adjust mininterval automatically
+                    instance.miniters = 1
+                    # Refresh now!
+                    instance.refresh()
+
+    def report(self):
+        # return self.is_alive()  # TODO: does not work...
+        return not self.was_killed
+
+
 class tqdm(object):
     """
     Decorate an iterable object, returning an iterator which acts exactly
     like the original iterable, but prints a dynamically updating
     progressbar every time a value is requested.
     """
+
+    monitor_interval = 10  # set to 0 to disable the thread
```

Comment:
Yes but a few operations every 10 seconds (by default, we can set the monitor_interval higher) has a negligible impact. I profiled and in fact the monitor has no effect that I could notice, it was only `self.miniters` that's producing a 10% overhead (but in exchange we have the guarantee that a tqdm bar never gets frozen, I think it's worth the price!).

Also this change should work with any submodule, that's why I thought it may be fit in the core.